decode.go

```diff
@@ -339,9 +444,22 @@ func (d *decoder) unmarshal(n *node, out reflect.Value) (good bool) {
 }
 
 func (d *decoder) document(n *node, out reflect.Value) (good bool) {
-	if len(n.children) == 1 {
-		d.doc = n
-		d.unmarshal(n.children[0], out)
+	if len(n.children) == 1 || len(n.children) == 0 {
+		if len(n.children) == 1 {
+			d.doc = n
+			d.unmarshal(n.children[0], out)
+		}
+
+		// If there is a PreDoc string, insert it as the first child
+		if _, ok := out.Interface().(MapSlice); len(strings.TrimSpace(n.value)) > 0 && ok {
+			slice := reflect.ValueOf(MapSlice{
+				MapItem{
+					Key:   PreDoc(n.value),
```

Comment:
The comment is of type String and the predoc is of type PreDoc (even though it is based on a string). I couldn't think of a way to distinguish a predoc from a normal comment otherwise...